das/aggregator.go

```diff
@@ -0,0 +1,218 @@
+	requestedCert, _, err := arbstate.DeserializeDASCertFrom(cert)
+	if err != nil {
+		return nil, err
+	}
+
+	// Cert is the aggregate cert, validate it against DAS public keys
+	var servicesThatSignedCert []serviceDetails
+	var pubKeys []blsSignatures.PublicKey
+	for _, d := range a.services {
+		if requestedCert.SignersMask&d.signerMask != 0 {
+			servicesThatSignedCert = append(servicesThatSignedCert, d)
+			pubKeys = append(pubKeys, d.pubKey)
+		}
+	}
+	signedBlob := serializeSignableFields(*requestedCert)
+	sigMatch, err := blsSignatures.VerifySignature(requestedCert.Sig, signedBlob, blsSignatures.AggregatePublicKeys(pubKeys))
+	if err != nil {
+		return nil, err
+	}
+	if !sigMatch {
+		return nil, errors.New("Signature of data in cert passed in doesn't match")
+	}
+
+	blobChan := make(chan []byte)
+	errorChan := make(chan error)
+	subCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for _, d := range servicesThatSignedCert {
+		go func(ctx context.Context, d serviceDetails) {
+			blob, err := d.service.Retrieve(ctx, cert)
+			if err != nil {
+				errorChan <- err
+				return
+			}
+			var blobHash [32]byte
+			copy(blobHash[:], crypto.Keccak256(blob))
+			if blobHash == requestedCert.DataHash {
+				blobChan <- blob
+			} else {
+				errorChan <- fmt.Errorf("DAS (mask %X) returned data that doesn't match requested hash!", d.signerMask)
+			}
+		}(subCtx, d)
+	}
+
+	errorCount := 0
+	var errorCollection []error
+	for errorCount < len(servicesThatSignedCert) {
+		select {
+		case blob := <-blobChan:
+			return blob, nil
+		case err = <-errorChan:
+			errorCollection = append(errorCollection, err)
+			log.Warn("Couldn't retrieve message from DAS", "err", err)
+			errorCount++
+		case <-ctx.Done():
+			break
+		}
+	}
+
+	return nil, fmt.Errorf("Data wasn't able to be retrieved from any DAS: %v", errorCollection)
+}
+
+type storeResponse struct {
+	cert *arbstate.DataAvailabilityCertificate
+	err  error
+
+	details serviceDetails
+}
+
+// Store calls Store on each backend DAS in parallel and collects responses; if
+// there were enough (at least K) responses then it aggregates the signatures and
+// signerMasks from each DAS together to put in the DataAvailabilityCertificate
+// that it returns; if it gets enough errors that K successes is impossible, then
+// it stops early and returns an error, and if it gets not enough successful
+// responses by the time its context is canceled (eg via DeadlineWrapper) then it
+// also returns an error.
+func (a *Aggregator) Store(ctx context.Context, message []byte, timeout uint64) (*arbstate.DataAvailabilityCertificate, error) {
+	var aggSignersMask uint64
+	var pubKeys []blsSignatures.PublicKey
+	var sigs []blsSignatures.Signature
+	var aggCert arbstate.DataAvailabilityCertificate
+
+	var initialStoreSucceeded bool
+	storeFailures := 0
+	if timeout == CALLEE_PICKS_TIMEOUT {
+		timeout = uint64(time.Now().Add(a.config.retentionPeriod).Unix())
+	}
+
+	responses := make(chan storeResponse)
```

Comment:
Let's make this channel buffered, so it can hold all of the possible responses; and `defer` closing it.

Reply:
I made it buffered with size `len(a.services)` so it can capture the response from all of the backend goroutines. As mentioned in the other channel related comment above I don't think we want to close it explicitly because there still could be backend goroutines running.